packages/grid/_modules_/grid/hooks/features/treeData/useGridTreeData.ts

```diff
@@ -121,24 +89,114 @@ export const useGridTreeData = (
     updateRowGrouping();
   }, [updateRowGrouping]);
 
+  /**
+   * PRE-PROCESSING
+   */
+  const getGroupingColDef = React.useCallback((): GridColDef => {
+    const propGroupingColDef = props.groupingColDef;
+
+    const baseColDef: GridColDef = {
+      ...GRID_TREE_DATA_GROUP_COL_DEF,
+      headerName: apiRef.current.getLocaleText('treeDataGroupingHeaderName'),
+      ...GRID_TREE_DATA_GROUP_COL_DEF_FORCED_PROPERTIES,
+    };
+    const colDefOverride: Partial<GridColDef> = propGroupingColDef ?? {};
+    // let colDefOverride: Partial<GridColDef>;
+    //
+    // if (typeof propGroupingColDef === 'function') {
+    //   const params: GridColDefOverrideParams = {
+    //     colDef: baseColDef,
+    //   };
+    //
+    //   colDefOverride = propGroupingColDef(params);
+    // } else {
+    //   colDefOverride = propGroupingColDef ?? {};
+    // }
+
+    return {
+      ...baseColDef,
+      ...colDefOverride,
+    };
+  }, [apiRef, props.groupingColDef]);
+
+  const updateGroupingColumn = React.useCallback(
+    (columnsState: GridColumnsRawState) => {
+      const groupingColDefField = GRID_TREE_DATA_GROUP_COL_DEF_FORCED_PROPERTIES.field;
+
+      const shouldHaveGroupingColumn = props.treeData;
+      const haveGroupingColumn = columnsState.lookup[groupingColDefField] != null;
```

Comment:
As written in my PR description, for the column pre-processing, the check to know if we need to do something has a complexity of `O(1)`.
The only linear complexity here is if we want to remove the grouping column (to find its index) but iit won't be need if the prep-processor is called after a change in a prop of another pre-processor.